### ISSUE TITLE
project status does not change to "Exporting"

### ISSUE THREAD

**Issue by amnonkhen** [ebi-ait/dcp-ingest-central#178]
This was noticed on 11/2 by @rays22, see #175.
It can be tracked to a deployment of the exporter from 29/1

**Comment by jacobwindsor** [comment]
After some investigation we found an issue with the state tracker in the logs: 

```
2021-01-29 17:41:18.846 ERROR 6 --- [           main] o.h.i.state.persistence.RedisPersister   : Failed to retrieve state machine with id: 7210fb46-add2-418f-905a-13f6d4f52753
com.esotericsoftware.kryo.KryoException: Unable to find class: org.humancellatlas.ingest.state.monitor.util.BundleTracker
        at com.esotericsoftware.kryo.util.DefaultClassResolver.readName(DefaultClassResolver.java:156) ~[kryo-shaded-3.0.3.jar!/:na]
        at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:133) ~[kryo-shaded-3.0.3.jar!/:na]
        at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:670) ~[kryo-shaded-3.0.3.jar!/:na]
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:781) ~[kryo-shaded-3.0.3.jar!/:na]
        at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:161) ~[kryo-shaded-3.0.3.jar!/:na]
        at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:39) ~[kryo-shaded-3.0.3.jar!/:na]
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:790) ~[kryo-shaded-3.0.3.jar!/:na]

...
```

It seems to be that my initial thoughts in [this PR](https://github.com/ebi-ait/ingest-core/pull/59?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDE1NzkwNzg2MDE6NTU3OTI3MA%3D%3D#pullrequestreview-589658803) are not the reason for the problem. Having said that, it is weird that we have `findByUuid` and `findByUuidUuid` endpoints for submissions when it is not the case for other metadata types - we should fix that but it is not urgent

**Comment by aaclan-ebi** [comment]
**Findings for today:**

There were errors when the state tracker was deployed 16 days ago, 29th January. The errors were not raised and were logged silently. And since the logs weren't checked, these errors were not discovered at that time.

There were 2 error types:
1. There were some connection errors in retrieving the submission envelope from core. The errors happened for Ray and Marion's submissions (cb156730-90b0-4b77-944c-bfc263204c61 , feebef33-476f-42e4-9bfa-6da47fb0d0f9 )

- The state tracker must have been deployed first before core and these could have caused those connection errors.
- We tried to fix connection errors by redeploying the state tracker.
- After redeploying the state tracker, we haven't seen any connection errors from before
    - There were many new http 404 errors  for submission uuids which are not in core. It's probably that those submissions envelopes are already deleted in core but state tracker still have them in its database.
- When Marion's submission feebef33-476f-42e4-9bfa-6da47fb0d0f9 was resubmitted, the state tracker still won't allow the submission to be transitioned from Valid to Submitted
        - Even though the state tracker initialisation seems to have been fixed. It seems that the restored state machines from the state tracker is out of sync with the actual state in Core database. This needs further investigation and fix.
	- We tried a workaround to set a metadata in that submission in back to Draft and that fixed the state machine for that submission. We tried to resubmit the submission and the submission and it was transitioned to Submitted -> Exporting -> Exported.
 
2. There were also 2 errors about missing BundleTracker class.
    - We didn't investigate this as this is not affecting the submission envelopes we want to export


**TLDR:**
1. There were issues in restoring the state machines in the state tracker and the state tracker database is out of sync from Ingest.
We need to investigate further why the data is out of sync and we need to sync the data.

2. There's a workaround for valid submissions. We could set a metadata in that submission back to Draft and wait for submission go to Valid again before exporting.

**Comment by aaclan-ebi** [comment]
I was able to setup the state tracker, core, validator (broker, ui for easier testing) locally and was able to debug and replicate the issue. 

Here are some of the things i've found/confirmed:

1. There's a circular dependency between state tracker and core. 
- Core is using http to send messages to the state tracker and would need the ip/url of the service.
- State tracker queries core with some info from submission envelope using uuid to load the data during its initialisation

- There is no issue in deploying the state tracker after core atm. The state tracker service ip seems to not change within the cluster, but i think this is not guaranteed.
- We probably would want to change core -> state tracker sending of messages so that it wouldn't need the state tracker service/endpoint. It used to be via RabbitMQ topic exchange, but the order of messages wasn't guaranteed it was quickly changed to be http and it's when the circular dependency was introduced. I believe we should be able to use RabbitMQ direct exchange if we want to make the order of messages guaranteed. This would also make the sending of events between our internal components consistent.

2. State tracker’s state machines are being persisted every minute. If there are errors in the loading of state machines during its initialisation/deployment it's going to get persisted. This is the most likely the reason why the state tracker data is corrupted/out of sync from ingest db. 

4. The data we store in Redis is the state machine context object and is not a human readable format. If we're going to correct the data we could probably do it in the code during the loading/initialisation.

2 actions:

1. (Urgent)Update the state tracker code to sync/correct the data during initialisation.
2. Remove the cyclic dependency.

**Comment by aaclan-ebi** [comment]
The AutloLoader in the state tracker ran successfully in dev:

[https://app.zenhub.com/files/299258791/357914b6-69a7-40da-992d-656f2997cb86/download](https://app.zenhub.com/files/299258791/357914b6-69a7-40da-992d-656f2997cb86/download)

The integration test is passing in dev:
https://gitlab.ebi.ac.uk/hca/ingest-integration-tests/-/jobs/371351

**Comment by aaclan-ebi** [comment]
PR is merged: https://github.com/ebi-ait/ingest-state-tracking/pull/4
The integration test is passing in staging https://gitlab.ebi.ac.uk/hca/ingest-integration-tests/-/pipelines/136211

**Comment by aaclan-ebi** [comment]
Released to prod today: https://github.com/ebi-ait/ingest-kube-deployment/blob/master/production/changelog.md#29-january-2021

https://embl-ebi-ait.slack.com/archives/CSDN3M2P5/p1615375886015000
